File: client/src/app/services/publish.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { interval, Observable, switchMap, takeWhile, tap } from 'rxjs';


export interface GetAdsRequest {
  page: number;
  size: number;
  sortBy?: string;
  sortOrder?: string;
  status?: 'PENDING_APPROVAL' | 'APPROVED' | 'REJECTED' | 'PUBLISHED';
}

export function getPossibleStatuses(): GetAdsRequest['status'][] {
  return ['PENDING_APPROVAL', 'APPROVED', 'REJECTED', 'PUBLISHED'];
}

export interface AdPaginatedResponse {
  content: AdResponse[];
  currentPage: number;
  totalPages: number;
  totalItems: number;
}

export interface AdResponse {
  id: number;
  title: string;
  description: string;
  squareMeters: number;
  price: number;
  location: string;
  latitude: number;
  longitude: number;
  status: string;
  accountId: number;
}

export interface AdStatusResponse {
  status: string;
}

export interface PublishGetAdsResponse {
  adsResponse: AdResponse[];
}

export interface PublishAdRejectRequest {
  reason: string;
}

export interface OpenPublishAccountRequest {
  username: string;
  password: string;
  email: string;
  name: string;
  surname: string;
  mobile: string;
}

export interface OpenPublishAccountResponse {
  id: number;
  username: string;
  walletId: string; // UUID as string
}

export interface PublishAdRequest {
  accountId: number | null;
  latitude: number | null;
  location: string |null;
  longitude: number | null;
  price: number | null;
  squareMeters: number | null;
  title: string | null;
  description?: string |null;
}

export interface PublishAdResponse {
  adId: number; // int64
  title: string;
}




@Injectable({
  providedIn: 'root'
})
export class PublishService {
  private baseUrl = '/api/publish';

  constructor(private http: HttpClient) {}

  /**
   * Approve an ad.
   * @param adId - ID of the ad to approve.
   */
  approveAd(adId: number): Observable<any> {
    const url = `${this.baseUrl}/ad/${adId}/approve`;
    const headers = new HttpHeaders({ 
      'Content-Type': 'application/json',
      'Use-Auth': 'true' // tells interceptor to include bearer token
    });
    return this.http.put<any>(url, null, { headers });
  }

  /**
   * Publish a new ad.
   * @param publishAdRequest - Object containing ad details.
   */
  publishAd(publishAdRequest: PublishAdRequest): Observable<PublishAdResponse> {
    const url = `${this.baseUrl}/ad`;
    const headers = new HttpHeaders({ 
      'Content-Type': 'application/json',
      'Use-Auth': 'true' // tells interceptor to include bearer token
    });
    return this.http.post<PublishAdResponse>(url, publishAdRequest, { headers });
  }

  /**
   * Get the status of an ad.
   * @param adId 
   * @returns 
   */
  getAdStatus(adId: number): Observable<AdStatusResponse> {
    const headers = new HttpHeaders({ 
      'Content-Type': 'application/json',
      'Use-Auth': 'true' // tells interceptor to include bearer token
    });
    return this.http.get<AdStatusResponse>(`${this.baseUrl}/ad/${adId}/status`, { headers });
  }

  /**
   * Poll the status of an ad.
   * @param adId - ID of the ad to poll.
   * @param pollInterval - Interval in milliseconds between polls.
   * @returns Observable that emits the status of the ad.
   *         Completes when the status is no longer 'PENDING_APPROVAL'.
   *        Errors if the ad does not exist.
   *       Unsubscribing stops the polling.
   *     The first value emitted is the current status of the ad.
   */
  pollStatus(adId: number, pollInterval : number = 2000): Observable<AdStatusResponse> {
    return interval(pollInterval).pipe( // Polling ogni 2 secondi
      switchMap(() => this.getAdStatus(adId)),
      takeWhile(response => response.status === 'PENDING_APPROVAL', true), // Ferma il polling quando lo stato cambia
      tap(response => console.log('Stato attuale:', response.status))
    );
  }


  /**
   * Reject an ad.
   * @param adId - ID of the ad to reject.
   * @param rejectionReason - Reason for rejection.
   */
  rejectAd(adId: number, rejectionReason: PublishAdRejectRequest ): Observable<any> {
    const url = `${this.baseUrl}/ad/${adId}/reject`;
    const headers = new HttpHeaders({ 
      'Content-Type': 'application/json',
      'Use-Auth': 'true' // tells interceptor to include bearer token
    });
    return this.http.put<any>(url, rejectionReason, { headers });
  }

  getAds(request: GetAdsRequest): Observable<AdPaginatedResponse> {
    const headers = new HttpHeaders({ 
      'Content-Type': 'application/json',
      'Use-Auth': 'true' // tells interceptor to include bearer token
    });
    return this.http.post<AdPaginatedResponse>(
      `${this.baseUrl}/ads`,
      request,
      { headers }
    );
  }

  getAdsAccount(accountId: number): Observable<PublishGetAdsResponse> {
    const headers = new HttpHeaders({ 
      'Content-Type': 'application/json',
      'Use-Auth': 'true' // tells interceptor to include bearer token
    });
    return this.http.get<PublishGetAdsResponse>(
      `${this.baseUrl}/ads/account/${accountId}`,
      { headers }
    );
  }

  /**
   * Open a new publish account.
   * @param openPublishAccountRequest - Object containing account details.
   */
  openPublishAccount(openPublishAccountRequest: OpenPublishAccountRequest): Observable<OpenPublishAccountResponse> {
    const url = `${this.baseUrl}/account`;
    const headers = new HttpHeaders({
      'Content-Type': 'application/json',
      'Use-Auth': 'true' // tells interceptor to include bearer token
    });
    return this.http.post<OpenPublishAccountResponse>(url, openPublishAccountRequest, { headers });
  }
}

File: publish-service/src/main/java/it/disim/univaq/sose/publishservice/service/PublishService.java
package it.disim.univaq.sose.publishservice.service;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.parameters.RequestBody;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import it.disim.univaq.sose.publishservice.domain.dto.*;
import jakarta.ws.rs.*;
import jakarta.ws.rs.container.AsyncResponse;
import jakarta.ws.rs.container.Suspended;


/**
 * The PublishService interface defines the service methods for the publish service.
 */
@Path("/api/publish")
public interface PublishService {
    /**
     * Publish an ad.
     *
     * @param ad the ad to publish
     * @param asyncResponse the async response
     * @return the response
     */
    @Operation(
            operationId = "publishAd",
            description = "Publish an ad.",
            responses = {
                    @ApiResponse(
                            responseCode = "200",
                            description = "Ad published successfully",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = PublishAdResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = PublishAdResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "400",
                            description = "Bad Request",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "415",
                            description = "Bad Requestss",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "503",
                            description = "Service Unavailable",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "500",
                            description = "Internal Server Error",
                            content = {
                                    @Content(
                                            mediaType = "application/json",
                                            schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(
                                            mediaType = "application/xml",
                                            schema = @Schema(implementation = ErrorResponse.class))
                            })
            }
    )
    @POST
    @Path("/ad")
    @Consumes({"application/json", "application/xml"})
    @Produces({"application/json", "application/xml"})
    void publishAd(@RequestBody(
            description = "PublishAdRequest",
            required = true,
            content = {
                    @Content(mediaType = "application/json", schema = @Schema(implementation = PublishAdRequest.class)),
                    @Content(mediaType = "application/xml", schema = @Schema(implementation = PublishAdRequest.class))}
    ) PublishAdRequest ad, @Suspended AsyncResponse asyncResponse);


    /**
     * Approve an ad.
     *
     * @param adId the ad ID
     * @param asyncResponse the async response
     * @return the response
     */
    @Operation(
            operationId = "approveAd",
            description = "Approve an ad.",
            responses = {
                    @ApiResponse(
                            responseCode = "200",
                            description = "Ad approved successfully",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = PublishAdApproveResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = PublishAdApproveResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "400",
                            description = "Bad Request",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "415",
                            description = "Bad Request",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "503",
                            description = "Service Unavailable",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "500",
                            description = "Internal Server Error",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            })
            }
    )
    @PUT
    @Consumes({"application/json", "application/xml"})
    @Produces({"application/json", "application/xml"})
    @Path("/ad/{adId}/approve")
    void approveAd(@PathParam("adId") long adId , @Suspended AsyncResponse asyncResponse);



    /**
     * Reject an ad.
     *
     * @param adId the ad ID
     * @param asyncResponse the async response
     * @param rejectRequest the reject request
     * @return the response
     */
    @Operation(
            operationId = "rejectAd",
            description = "Reject an ad.",
            responses = {
                    @ApiResponse(
                            responseCode = "200",
                            description = "Ad rejected successfully",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = PublishAdRejectResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = PublishAdRejectResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "400",
                            description = "Bad Request",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "415",
                            description = "Bad Request",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "503",
                            description = "Service Unavailable",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "500",
                            description = "Internal Server Error",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            })
            }
    )
    @PUT
    @Consumes({"application/json", "application/xml"})
    @Produces({"application/json", "application/xml"})
    @Path("/ad/{adId}/reject")
    void rejectAd(@PathParam("adId") long adId,
                  @RequestBody(
            description = "Reason for rejection",
            required = true,
            content = {
                    @Content(mediaType = "application/json", schema = @Schema(implementation = PublishAdRejectRequest.class)),
                    @Content(mediaType = "application/xml", schema = @Schema(implementation = PublishAdRejectRequest.class))}
    ) PublishAdRejectRequest rejectRequest, @Suspended AsyncResponse asyncResponse);

    /**
     * Get all ads.
     *
     * @param asyncResponse the async response
     * @return the response
     */
    @Operation(
            operationId = "getAds",
            description = "Get all ads.",
            responses = {
                    @ApiResponse(
                            responseCode = "200",
                            description = "Ads retrieved successfully",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = AdPaginatedResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = AdPaginatedResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "400",
                            description = "Bad Request",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "415",
                            description = "Bad Request",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "503",
                            description = "Service Unavailable",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "500",
                            description = "Internal Server Error",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            })
            }
    )
    @POST
    @Consumes({"application/json", "application/xml"})
    @Produces({"application/json", "application/xml"})
    @Path("/ads")
    void getAds(@RequestBody(
            description = "GetAdsRequest",
            required = true,
            content = {
                    @Content(mediaType = "application/json", schema = @Schema(implementation = GetAdsRequest.class)),
                    @Content(mediaType = "application/xml", schema = @Schema(implementation = GetAdsRequest.class))
            }
    ) GetAdsRequest getAdsRequest, @Suspended AsyncResponse asyncResponse);


    /**
     * Get all ads by account ID.
     *
     * @param accountId the account ID
     * @param asyncResponse the async response
     * @return the response
     */
    @Operation(
            operationId = "getAdsByAccountId",
            description = "Get all ads by account ID.",
            responses = {
                    @ApiResponse(
                            responseCode = "200",
                            description = "Ads retrieved successfully",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = PublishGetAdsResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = PublishGetAdsResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "400",
                            description = "Bad Request",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "415",
                            description = "Bad Request",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "503",
                            description = "Service Unavailable",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "500",
                            description = "Internal Server Error",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            })
            }
    )
    @GET
    @Produces({"application/json", "application/xml"})
    @Path("/ads/account/{accountId}")
    void getAdsByAccountId(@PathParam("accountId") long accountId, @Suspended AsyncResponse asyncResponse);


    /**
     * Open a publish account.
     *
     * @param openPublishAccountRequest the open publish account request
     * @param asyncResponse the async response
     * @return the response
     */
    @Operation(
            operationId = "openPublishAccount",
            description = "Open a publish account.",
            responses = {
                    @ApiResponse(
                            responseCode = "200",
                            description = "Account opened successfully",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = OpenPublishAccountResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = OpenPublishAccountResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "400",
                            description = "Bad Request",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "415",
                            description = "Bad Request",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "503",
                            description = "Service Unavailable",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }),
                    @ApiResponse(
                            responseCode = "500",
                            description = "Internal Server Error",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            })
            }
    )
    @POST
    @Consumes({"application/json", "application/xml"})
    @Produces({"application/json", "application/xml"})
    @Path("/account")
    void openPublishAccount(@RequestBody(
            description = "OpenPublishAccountRequest",
            required = true,
            content = {
                    @Content(mediaType = "application/json", schema = @Schema(implementation = OpenPublishAccountRequest.class)),
                    @Content(mediaType = "application/xml", schema = @Schema(implementation = OpenPublishAccountRequest.class))
            }
    ) OpenPublishAccountRequest openPublishAccountRequest, @Suspended AsyncResponse asyncResponse);
}
